src/ray/gcs/gcs_server/gcs_resource_report_poller.h

```diff
@@ -0,0 +1,97 @@
+#include "ray/gcs/gcs_server/gcs_resource_manager.h"
+#include "ray/rpc/node_manager/node_manager_client_pool.h"
+
+namespace ray {
+namespace gcs {
+
+/// Polls raylets on a separate thread to update GCS's view of the cluster's resource
+/// utilization. This class creates and manages a polling thread. All public methods are
+/// thread safe.
+class GcsResourceReportPoller {
+  /*
+  This class roughly polls each node independently (with the exception of max
+  concurrency). The process for polling a single node is as follows:
+
+  A new node joins the cluster.
+  1. (Main thread) Begin tracking the node, and begin the polling process.
+
+  Main polling procedure.
+  2. (Polling thread) Enqueue the node to be pulled.
+  3. (Polling thread) Node is popped off the back of the queue and RequestResourceReport
+  is sent to the raylet.
+  4. (Main thread) The raylet responds and the resource manager is updated. This section
+  is _not_ thread safe (i.e. should not modify the resource report poller state).
+  5. (Polling thread) The RequestResourceReport continuation runs, scheduling the next
+  pull time.
+  6. (Polling thread) The next pull time occurs, and step 2 is repeated.
+
+  The node leaves the cluster.
+  7. Untrack the node. The next time the main polling procedure comes across the node, it
+  should be dropped from the system.
+   */
+
+ public:
+  GcsResourceReportPoller(uint64_t max_concurrent_pulls,
+                          std::shared_ptr<GcsResourceManager> gcs_resource_manager,
+                          std::shared_ptr<rpc::NodeManagerClientPool> raylet_client_pool);
+
+  ~GcsResourceReportPoller();
+
+  /// Start a thread to poll for resource updates.
+  void Start();
+
+  /// Stop polling for resource updates.
+  void Stop();
+
+  /// Event handler when a new node joins the cluster.
+  void HandleNodeAdded(std::shared_ptr<rpc::GcsNodeInfo> node_info);
+
+  /// Event handler when a node leaves the cluster.
+  void HandleNodeRemoved(std::shared_ptr<rpc::GcsNodeInfo> node_info);
```

Comment:
Nit: Use [thread-safety annotations](https://abseil.io/docs/cpp/guides/synchronization#thread-annotations).

```suggestion
  void HandleNodeRemoved(std::shared_ptr<rpc::GcsNodeInfo> node_info)
      LOCKS_EXCLUDED(mutex_);
```